### PR TITLE
Cohort analytics

### PR DESCRIPTION
**Pull request by neilisfragile**
Create a new table to store daily user visit information which can be used in cohort analysis.

Rather than insert on each user action, instead slurp the user_ips table daily.

Not sure how best to test this - any pointers?

### CODE REVIEW

**Comment by erikjohnston** [comment]
I would be tempted to work out the active users for today as you go rather than yesterday. I.e. every hour running something like:

```sql
INSERT INTO user_daily_visits (user_id, device_id, timestamp)
SELECT user_id, device_id, $today
FROM user_ips AS u
LEFT JOIN user_daily_visits USING (user_id, device_id)
WHERE last_seen > $today AND timestamp IS NULL
```

(The join to ensure that we don't insert duplicate values)

**Comment by neilisfragile** [comment]
Note change in SQL to handle case of duplicates on insert (where there are multiple entries that did not match previous entries in user_daily_visits)

File: synapse/storage/__init__.py [hunk 1]
```diff
@@ -347,6 +351,62 @@ def _count_r30_users(txn):
...
+            txn.execute(sql, (today_start, today_start,
+                              self._last_user_visit_update,
+                              today_start + a_day_in_milliseconds))
+            # Update _last_user_visit_update to now. The reason to do this
+            # rather just clamping to the beginning of the day is to limit
+            # the size of the join - meaning that the query can be run more
+            # frequently
+
+            now = datetime.datetime.utcnow()
+            self._last_user_visit_update = int(time.mktime(now.timetuple())) * 1000
```
Comment (erikjohnston) [inline synapse/storage/__init__.py]: You can just use `self.clock.time_msec()`

Comment (erikjohnston) [reply synapse/storage/__init__.py]: In fact, You probably want to get this before the query, so you do:

```python
now = self.clock.time_msec()
txn.execute(sql, (
    today_start, today_start,
    self._last_user_visit_update,
    now,
))
self._last_user_visit_update = now
```

This ensures there isn't going to be any overlaps or missed updates.

File: synapse/storage/__init__.py [hunk 2]
```diff
@@ -347,6 +351,62 @@ def _count_r30_users(txn):
...
+                    ON u.user_id = udv.user_id AND u.device_id=udv.device_id
+                    WHERE last_seen > ? AND last_seen <= ? AND udv.timestamp IS NULL
+            """
+
+            # This means that the day has rolled over but there could still
+            # be entries from the previous day. There is an edge case
+            # where if the user logs in at 23:59 and overwrites their
+            # last_seen at 00:01 then they will not be counted in the
+            # previous day's stats - it is important that the query is run
+            # to minimise this case.
```
Comment (erikjohnston) [inline synapse/storage/__init__.py]: Do you mean:

> it is important that the query is run ***often***...

File: synapse/storage/__init__.py [hunk 3]
```diff
@@ -347,6 +351,62 @@ def _count_r30_users(txn):
...
+            if today_start > self._last_user_visit_update:
+                yesterday_start = today_start - a_day_in_milliseconds
+                txn.execute(sql, (yesterday_start, yesterday_start,
+                                  self._last_user_visit_update, today_start))
+                self._last_user_visit_update = today_start
+
+            txn.execute(sql, (today_start, today_start,
+                              self._last_user_visit_update,
+                              today_start + a_day_in_milliseconds))
```
Comment (erikjohnston) [inline synapse/storage/__init__.py]: (The identing style we use tends to not use this continuation indent. See other comment for how it normally looks)

**Pull request merged**